asciidtype/asciidtype/src/casts.c

```diff
@@ -72,33 +76,243 @@ ascii_to_ascii(PyArrayMethod_Context *context, char *const data[],
+                        PyExc_TypeError,
+                        "Can only store ASCII text in a ASCIIDType array.");
+                return -1;
+            }
+            // UCS4 character is ascii, so copy first byte of character
+            // into output, ignoring the rest
+            *(out + i) = *(in + i * 4);
+        }
+        // write zeros to remaining ASCII characters (if any)
+        for (int i = copy_size; i < out_size; i++) {
+            *(out + i) = '\0';
+        }
+        in += in_stride;
+        out += out_stride;
+    }
+
+    return 0;
+}
+
 static int
-ascii_to_ascii_get_loop(PyArrayMethod_Context *context, int aligned,
-                        int NPY_UNUSED(move_references),
-                        const npy_intp *strides,
-                        PyArrayMethod_StridedLoop **out_loop,
-                        NpyAuxData **NPY_UNUSED(out_transferdata),
-                        NPY_ARRAYMETHOD_FLAGS *flags)
+ascii_to_unicode(PyArrayMethod_Context *context, char *const data[],
+                 npy_intp const dimensions[], npy_intp const strides[],
+                 NpyAuxData *NPY_UNUSED(auxdata))
 {
-    *out_loop = (PyArrayMethod_StridedLoop *)&ascii_to_ascii;
+    PyArray_Descr **descrs = context->descriptors;
+    long in_size = ((ASCIIDTypeObject *)descrs[0])->size;
+    long out_size = (descrs[1]->elsize) / 4;
+    long copy_size;
 
-    *flags = 0;
+    if (out_size > in_size) {
+        copy_size = in_size;
+    }
+    else {
+        copy_size = out_size;
+    }
+
+    npy_intp N = dimensions[0];
+    char *in = data[0];
+    char *out = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+
+    while (N--) {
+        // copy ASCII input to first byte, fill rest with zeros
+        for (int i = 0; i < copy_size; i++) {
+            ((Py_UCS4 *)out)[i] = ((Py_UCS1 *)in)[i];
+        }
+        // fill all remaining UCS4 characters with zeros
+        for (int i = copy_size; i < out_size; i++) {
+            ((Py_UCS4 *)out)[i] = (Py_UCS1)0;
+        }
+        in += in_stride;
+        out += out_stride;
+    }
     return 0;
 }
 
+static NPY_CASTING
+ascii_to_unicode_resolve_descriptors(PyObject *NPY_UNUSED(self),
+                                     PyArray_DTypeMeta *NPY_UNUSED(dtypes[2]),
+                                     PyArray_Descr *given_descrs[2],
+                                     PyArray_Descr *loop_descrs[2],
+                                     npy_intp *NPY_UNUSED(view_offset))
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+    long in_size = ((ASCIIDTypeObject *)given_descrs[0])->size;
+    if (given_descrs[1] == NULL) {
+        PyArray_Descr *unicode_descr = PyArray_DescrNewFromType(NPY_UNICODE);
+        // numpy stores unicode as UCS4 (4 bytes wide), so bitshift
+        // by 2 to get the number of bytes needed to store the UCS4 charaters
+        unicode_descr->elsize = in_size << 2;
```

Comment:
My very personal thing would be to use `* 4` and `/ 4` and trust the compiler to know that it's just a bit shift.  (not that I checked that they do :)).